### PR TITLE
Rental bikes on Point Verdant

### PR DESCRIPTION
**Pull request by DreamySkrell**
image
image
image
image

- needs https://github.com/Aurorastation/Aurora.3/pull/18457 to be merged first
- can "buy" a bike key - actually rent it for a day
- 50cr for a moped, 200cr for a sports bike
- also maps in non-rental bikes and bike keys around, like for the police or gangsters

### CODE REVIEW

**Comment by DreamySkrell** [comment]
!wip

**Comment by GMR25** [comment]
Is there any risk of shenanigans if the round occurs during an IC day change? It might be helped by a receipt that includes the time of purchase and ID of the buyer.

**Comment by DreamySkrell** [comment]
> Is there any risk of shenanigans if the round occurs during an IC day change? It might be helped by a receipt that includes the time of purchase and ID of the buyer.

image

it expires in 24 hours from purchase

**Comment by DreamySkrell** [comment]
!review

File: code/modules/modular_computers/hardware/hard_drive.dm [hunk 1]
```diff
@@ -156,6 +156,8 @@
 /obj/item/computer_hardware/hard_drive/Destroy()
 	if(parent_computer?.hard_drive == src)
 		parent_computer.hard_drive = null
+	for(var/datum/computer_file/file in stored_files)
+		file.hard_drive = null
```
Comment (FluffyGhoster) [inline code/modules/modular_computers/hardware/hard_drive.dm:160]: This isn't correct, if the hard drive is deleted, the programs it contains should be cleanly deleted too, not left hanging unlinked

**Review by FluffyGhoster — CHANGES_REQUESTED** [review]

File: code/game/machinery/vending.dm [hunk 2]
```diff
@@ -168,6 +169,14 @@
 	add_overlay(screen_overlays["[icon_state]-[deny ? "deny" : "screen"]"])
 	reset_light()
 
+/**
+ *  Build src.products
+ *
+ *  To be overriden if building the products list is dynamic at runtime or needs any complex logic
+ */
+/obj/machinery/vending/proc/build_products()
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending.dm:177]: should not sleep
doesn't have to be pure

File: code/game/machinery/vending.dm [hunk 3]
```diff
@@ -735,6 +745,10 @@
 				if(1)
 					use_power_oneoff(RC.reagents.set_temperature(heating_temperature))
 
+/// To be overriden if vended out products need any post-processing.
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending.dm:748]: done

File: code/game/machinery/vending.dm [hunk 4]
```diff
@@ -735,6 +745,10 @@
...
+/obj/machinery/vending/proc/vended_product_post(var/obj/vended)
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending.dm:749]: done

File: code/game/machinery/vending_types.dm [hunk 5]
```diff
@@ -1638,3 +1638,47 @@
...
+	)
+	restock_items = FALSE
+	random_itemcount = FALSE
+	light_color = COLOR_BABY_BLUE
+	/// List of strings.
+	/// Vended out keys will be filled with these key data (== bike reg plates) strings.
+	/// Also based on this list is filled the products assoc list.
+	var/list/key_data_mopeds = list()
+	/// Same as the list for mopeds, except for sports bikes.
+	var/list/key_data_sports = list()
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending_types.dm:1664]: done

File: code/game/machinery/vending_types.dm [hunk 6]
```diff
@@ -1638,3 +1638,47 @@
...
+
+/obj/machinery/vending/rental_bikes/build_products()
+	products[/obj/item/key/bike/moped] = key_data_mopeds.len
+	products[/obj/item/key/bike/sport] = key_data_sports.len
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending_types.dm:1668]: done

File: code/game/machinery/vending_types.dm [hunk 7]
```diff
@@ -1638,3 +1638,47 @@
...
+
+/obj/machinery/vending/rental_bikes/vended_product_post(var/obj/vended)
+	var/obj/item/key/key = vended
+	if(!istype(key))
+		return
+	// expires the next day
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending_types.dm:1674]: done

File: code/game/machinery/vending_types.dm [hunk 8]
```diff
@@ -1638,3 +1638,47 @@
...
+	var/rental_expiry = "[GLOB.game_year]-[time2text(world.realtime + 1 DAY, "MM-DD")] [worldtime2text()]"
+	key.desc += " Property of Idris Incorporated. Rental expires on [rental_expiry]. Return fully charged."
+	if(key_data_mopeds && istype(key, /obj/item/key/bike/moped))
```
[the remark this answers is not in the document]
Comment (DreamySkrell) [reply code/game/machinery/vending_types.dm:1677]: done

**Review by NonQueueingMatt — APPROVED** [review]

**Review by FluffyGhoster — APPROVED** [review]

**Review by Arrow768 — APPROVED** [review]

**Pull request merged**